nft/ethereum_payment_obligation.go

```diff
@@ -80,45 +88,155 @@ func newEthereumPaymentObligation(
+	if err != nil {
+		return mreq, err
+	}
+
+	var pfs []*proofspb.Proof
+	pfs = append(pfs, docProofs.FieldProofs...)
+	for _, p := range nftProofs {
+		p := p
+		pfs = append(pfs, &p)
 	}
 
-	toAddress := common.HexToAddress(depositAddress)
+	toAddress := common.HexToAddress(req.DepositAddress)
 
 	anchorID, err := anchors.ToAnchorID(corDoc.CurrentVersion)
 	if err != nil {
-		return MintRequest{}, err
+		return mreq, err
 	}
 
 	rootHash, err := anchors.ToDocumentRoot(corDoc.DocumentRoot)
 	if err != nil {
-		return MintRequest{}, err
+		return mreq, err
 	}
 
-	requestData, err := NewMintRequest(tokenID, toAddress, anchorID, proofs.FieldProofs, rootHash)
+	requestData, err := NewMintRequest(tokenID, toAddress, anchorID, pfs, rootHash)
 	if err != nil {
-		return MintRequest{}, err
+		return mreq, err
 	}
 
 	return requestData, nil
 
 }
 
+// generateNFTProofs generates required NFT uniqueness proofs
+func generateNFTProofs(cd *coredocumentpb.CoreDocument, req MintNFTRequest, tokenID TokenID, cid identity.CentID) ([]proofspb.Proof, error) {
+	cdTree, err := coredocument.GetCoreDocTree(cd)
+	if err != nil {
+		return nil, err
+	}
+
+	var pfs []proofspb.Proof
+	registry := common.HexToAddress(req.RegistryAddress)
+	if req.SubmitTokenProof {
+		pf, err := createTokenProof(cd, cdTree, registry)
+		if err != nil {
+			return nil, err
+		}
+
+		pfs = append(pfs, pf)
+	}
+
+	if req.GrantNFTReadAccess && req.SubmitNFTReadAccessProof {
+		pf, err := createNFTReadAccessProof(cd, cdTree, registry, tokenID[:])
+		if err != nil {
+			return nil, err
+		}
+
+		pfs = append(pfs, pf...)
+	}
+
+	if req.SubmitRoleProof != "" {
+		pf, err := createRoleProof(cd, cdTree, req.SubmitRoleProof, cid)
+		if err != nil {
+			return nil, err
+		}
+
+		pfs = append(pfs, pf)
+	}
+
+	return pfs, nil
+}
+
+func createRoleProof(cd *coredocumentpb.CoreDocument, cdTree *proofs.DocumentTree, roleKey string, cid identity.CentID) (proof proofspb.Proof, err error) {
+	idx, role := getRoleForAccount(cd, roleKey, cid)
+	if role == nil {
+		return proof, ErrNFTRoleMissing
+	}
+
+	pk := fmt.Sprintf("roles[%s].collaborators[%d]", hexutil.Encode(role.RoleKey), idx)
+	return cdTree.CreateProof(pk)
+}
+
+// createNFTReadAccessProof creates proof that nft exists in the role
+func createNFTReadAccessProof(cd *coredocumentpb.CoreDocument, cdTree *proofs.DocumentTree, registry common.Address, tokenID []byte) (proofs []proofspb.Proof, err error) {
+
+	var rridx, ridx, nftidx int
+	var rk []byte
+	found := coredocument.FindRole(cd, coredocumentpb.Action_ACTION_READ, func(i, j int, role *coredocumentpb.Role) bool {
+		si, found := coredocument.IsNFTInRole(role, registry, tokenID)
+		if found {
+			rridx = i
+			ridx = j
+			rk = role.RoleKey
+			nftidx = si
+		}
+
+		return found
+	})
+
+	if !found {
+		return nil, ErrNFTRoleMissing
+	}
+
+	fields := []string{
+		fmt.Sprintf("read_rules[%d].roles[%d]", rridx, ridx),          // proof that a read rule exists with the nft role
```

Comment:
rridx & ridx also need to be hex encoded AFAIK

Reply:
Nope, else tests would have failed ;)

Reply:
hm, ok that's odd. We should definitely investigate that there are no bugs hidden somewhere there. I'll look into this.